v2/tools/asoctl/internal/crd/cleaner.go

```diff
@@ -0,0 +1,194 @@
+			klog.Infof("starting cleanup for '%s'", crd.Name)
+			objectsToMigrate, err := c.getObjectsForMigration(ctx, crd, deprecatedVersionRegexp)
+			if err != nil {
+				return err
+			}
+
+			err = c.migrateObjects(ctx, objectsToMigrate)
+			if err != nil {
+				return err
+			}
+
+			err = c.updateStorageVersions(ctx, crd, newStoredVersions)
+			if err != nil {
+				return err
+			}
+
+			updated++
+		} else {
+			klog.Infof("nothing to update for '%s'\n", crd.Name)
+		}
+	}
+
+	if !c.dryRun {
+		klog.Infof("updated %d CRD(s)\n", updated)
+
+	}
+
+	return nil
+}
+
+func (c *Cleaner) updateStorageVersions(
+	ctx context.Context,
+	crd apiextensions.CustomResourceDefinition,
+	newStoredVersions []string) error {
+
+	if c.dryRun {
+		klog.Infof("new storeVersions for '%s' CRD status storedVersions: %s\n", crd.Name, newStoredVersions)
+		return nil
+	}
+
+	crd.Status.StoredVersions = newStoredVersions
+	updatedCrd, err := c.apiExtensionsClient.UpdateStatus(ctx, &crd, v1.UpdateOptions{})
+	if err != nil {
+		return err
+	}
+	klog.Infof("updated '%s' CRD status storedVersions to : %s\n", crd.Name, updatedCrd.Status.StoredVersions)
+
+	return nil
+}
+
+func (c *Cleaner) migrateObjects(ctx context.Context, objectsToMigrate *unstructured.UnstructuredList) error {
+	for _, obj := range objectsToMigrate.Items {
+		obj := obj
+		if c.dryRun {
+			// TODO: verbose
+			klog.V(logging.Verbose).Infof("resource '%s' to migrate for kind '%s'", obj.GetName(), obj.GroupVersionKind().Kind)
+			continue
+		}
+
+		err := retry.OnError(c.migrationBackoff, isErrorFatal, func() error { return c.client.Update(ctx, &obj) })
+		if isErrorFatal(err) {
+			return err
+		}
+		klog.V(logging.Verbose).Infof("migrated '%s' for %s\n", obj.GetName(), obj.GroupVersionKind().Kind)
+	}
+
+	return nil
+}
+
+func isErrorFatal(err error) bool {
+	if err == nil {
+		return false
+	}
+
+	if apierrors.IsGone(err) { // If resource no longer exists, we don't want to retry
+		return false
+	} else if apierrors.IsConflict(err) { // If resource is already in the state of update, we don't want to retry either
+		return false
+	} else {
+		return true
+	}
+}
+
+func (c *Cleaner) getObjectsForMigration(ctx context.Context, crd apiextensions.CustomResourceDefinition, versionRegexp *regexp.Regexp) (*unstructured.UnstructuredList, error) {
+	list := &unstructured.UnstructuredList{}
+
+	if ok, version := requireMigrate(crd, versionRegexp); ok {
+		list.SetGroupVersionKind(schema.GroupVersionKind{
+			Group:   crd.Spec.Group,
+			Version: version,
+			Kind:    crd.Spec.Names.ListKind,
+		})
+
+		if err := c.client.List(ctx, list); err != nil {
+			return nil, err
+		}
+	}
+
+	return list, nil
+}
+
+// requireMigrate checks if resources under CRD need migration. Requirement for migration would depend on if the deprecated version is the storage version.
+func requireMigrate(crd apiextensions.CustomResourceDefinition, versionRegexp *regexp.Regexp) (bool, string) {
```

Comment:
See my comment on your tests, I think you have this check backwards.

I also think it might make more sense to do this check in `removeMatchingStoredVersions` and have that function return an error if it attempts to remove a version that matches the regex but is also marked as the storage version.

So basically the flow would be:

1. List CRDs
2. For each crd...
    1. Remove the version in storedVersions that matches the pattern, as long as it's not marked the storageVersion
    2. ... Do everything you're already doing.

Then I think you could delete this method and its usage?

Reply:
This method would still be required to fetch a version we want to update right? Basically to get version for GVK of resources we want to list and migrate? which would be alpha version only I assume?